src/shipit_code_coverage/shipit_code_coverage/codecov.py

```diff
@@ -131,16 +131,25 @@ def download_coverage_artifacts(self):
 
         def should_download(status, chunk_name, platform_name):
             with downloaded_tasks_lock:
+                # If the chunk hasn't been downloaded before, this is obviously the best task
+                # to download it from.
                 if (chunk_name, platform_name) not in downloaded_tasks:
-                    return True
+                    download_lock = Lock()
+                    downloaded_tasks[(chunk_name, platform_name)] = {
+                        'status': status,
+                        'lock': download_lock,
+                    }
+                else:
+                    task = downloaded_tasks[(chunk_name, platform_name)]
 
-                other_status = downloaded_tasks[(chunk_name, platform_name)]
+                    if (status == 'failed' and task.status == 'exception') or (status == 'completed' and task.status != 'completed'):
```

Comment:
Nit: i would use an enum for the status here or simply declare the status as uppercases "const" in a unique file